scylla/src/transport/session.rs

```diff
@@ -1814,13 +1807,63 @@ impl Session {
         prepared: &PreparedStatement,
         serialized_values: &SerializedValues,
     ) -> Result<Option<Token>, QueryError> {
-        match self.calculate_partition_key(prepared, serialized_values) {
-            Ok(Some(partition_key)) => {
-                let partitioner_name = prepared.get_partitioner_name();
-                Ok(Some(partitioner_name.hash(&partition_key)))
-            }
-            Ok(None) => Ok(None),
-            Err(err) => Err(err),
+        Ok(self
+            .calculate_partition_key(prepared, serialized_values)?
+            .map(|partition_key| prepared.get_partitioner_name().hash(&partition_key)))
+    }
+
+    /// Get the first node/shard that the load balancer would target if running this query
+    ///
+    /// This may help constituting shard-aware batches
+    pub fn first_shard_for_statement(
+        &self,
+        prepared: &PreparedStatement,
+        serialized_values: &SerializedValues,
+    ) -> Result<Option<(Arc<Node>, Option<routing::Shard>)>, QueryError> {
```

Comment:
Naming nit: as this returns a node along with a shard, please mention it in the function's name, e.g. `first_node_and_shard_for_statement()`.

Reply:
The way I understand it a shard in a cluster is only entirely determined by the `(Node, ShardIdxOnNode)` tuple, which is what's returned here. It seems that it's rather `routing::Shard` that's poorly named, because it's named as if it uniquely determines a shard, but in fact it's only a `ShardIdxOnNode` `u32` and it needs to be associated with its `Node` to do so.

Reply:
You're right, our naming here is inaccurate. It's probable that in a future release we will rename `Shard` to `ShardIdx` or something similar.